crates/executor/src/transaction.rs

```diff
@@ -73,3 +80,364 @@ pub fn gas_vector_computation_mode(transaction: &Transaction) -> GasVectorComput
+            // transaction to execute without running out of L2 gas.
+            loop {
+                tracing::debug!(
+                    "Searching for minimal L2 gas limit in range [{lower_bound}; {upper_bound}]. \
+                     Current limit: {current_l2_gas_limit}"
+                );
+                set_l2_gas_limit(tx, current_l2_gas_limit);
+
+                // Special case where the search would get stuck if `current_l2_gas_limit ==
+                // lower_bound` but the required amount is equal to the upper bound.
+                let bounds_diff = upper_bound
+                    .checked_sub(lower_bound)
+                    .expect("Upper bound >= lower bound");
+                if bounds_diff == GasAmount(1) && current_l2_gas_limit == lower_bound {
+                    lower_bound = upper_bound;
+                    current_l2_gas_limit = upper_bound;
+                }
+
+                match simulate_transaction(tx, tx_index, state, block_context) {
+                    Ok(_) => {
+                        if search_done(lower_bound, upper_bound, L2_GAS_SEARCH_MARGIN) {
+                            break;
+                        }
+
+                        upper_bound = current_l2_gas_limit;
+                        current_l2_gas_limit = midpoint(lower_bound, upper_bound);
+                    }
+                    Err(TransactionSimulationError::OutOfGas) => {
+                        lower_bound = current_l2_gas_limit;
+                        current_l2_gas_limit = midpoint(lower_bound, upper_bound);
+                    }
+                    Err(TransactionSimulationError::ExecutionError(error)) => {
+                        return Err(error);
+                    }
+                }
+            }
+
+            current_l2_gas_limit
+        }
+        Err(TransactionSimulationError::ExecutionError(error)) => {
+            return Err(error);
+        }
+    };
+
+    if l2_gas_limit > initial_l2_gas_limit {
+        tracing::debug!(
+            initial_limit=%initial_l2_gas_limit,
+            final_limit=%l2_gas_limit,
+            "Initial L2 gas limit exceeded"
+        );
+        // Set the L2 gas limit to zero so that the transaction reverts with a detailed
+        // `ExecutionError`.
+        set_l2_gas_limit(tx, GasAmount::ZERO);
+        match execute_transaction(tx, tx_index, state, block_context) {
+            Err(e @ TransactionExecutionError::ExecutionError { .. }) => {
+                return Err(e);
+            }
+            _ => unreachable!("Transaction should revert when gas limit is zero"),
+        }
+    }
+
+    // Finally, execute the transaction with the found L2 gas limit and set that
+    // limit as the estimate.
+    set_l2_gas_limit(tx, l2_gas_limit);
+    let mut tx_info = execute_transaction(tx, tx_index, state, block_context)
+        .expect("Transaction already executed successfully");
+    tx_info.receipt.gas.l2_gas = l2_gas_limit;
```

Comment:
Could this execution be optimized away?

Seems to be the same as the last `simulate_transaction()` call (except that simulate does not persist the state).